### PR TITLE
Implemented CS Status Request

### PR DESCRIPTION
**Pull request by mtraneis**
Things to note:
Used freemarker to implement the output message. 
Created a utility class "FreemarkerUtils.java" to transform data into SIP messages.

Note: The enumeration messages were automatically formatted. I'm guessing that it was due to the Checkstyle?  Basically there's a space here or there, and there shouldn't be a semicolon after the last element in the enum.

### CODE REVIEW

**Comment by mreno-EBSCO** [comment]
The semicolon is out of habit. It is required if there is any logic besides the enum fields, like a constructor or methods, so I find it more consistent. The Google style guide has enumeration examples without the semicolon, but does not explicitly forbid the semicolon. Checkstyle shouldn't automatically do anything, so I'm not sure what you mean there.

**Comment by mtraneis** [comment]
> The semicolon is out of habit. It is required if there is any logic besides the enum fields, like a constructor or methods, so I find it more consistent. The Google style guide has enumeration examples without the semicolon, but does not explicitly forbid the semicolon. Checkstyle shouldn't automatically do anything, so I'm not sure what you mean there.

Maybe my IDE automatically formatted it to whatever setting it thought was awesome. I didn't consciously touch any of these files. Do you want me to revert them back?

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  /**
+   * Utility method that executes a freemarker template to return a templated-formmated string.
+   *
+   * @param data the data used by the template.
+   * @param template the template to apply to the data.
+   * @return String output of running the template on the data
+   */
+  public static String executeFreemarkerTemplate(Object data, Template template) {
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java]: Maybe something to think about later, but we may want to decouple Freemarker with the actual business logic for building messages. Maybe have a "templating" abstraction and we happen to use Freemarker as the implementation, but we could replace it with some other templating system without changing the message building code.

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    } catch (TemplateException e) {
+      log.error("Having problems finding and loading template: " + e.getMessage());
+    } catch (IOException ioEx) {
+      log.error("Having problems applying template to data: " + ioEx.getMessage());
+    } catch (Exception ex) {
+      log.error("Error applying template to data: " + ex.getMessage());
+    }
+
+    log.debug("Data =" + (data == null ? "" : data.toString())
+              + "Template =" + (template == null ? "" : template.getName()));
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java]: We might want to build this via the logger, so we don't construct a string that will get thrown away when logging is not enabled,  something like:
```suggestion
    log.debug("Data = {} Template = {}",  data == null ? "" : data.toString(), template == null ? "" : template.getName());
```
Or better yet:
```suggestion
    log.debug("Data = {} Template = {}", () -> {data == null ? "" : data.toString()}, () -> {template == null ? "" : template.getName()});
```
With this approach, `data.toString()`s and the `getName()` are not called unless debug is executed.

File: src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java [hunk 3]
```diff
@@ -79,8 +81,38 @@ public void canStartMainVericleInjectingSip2RequestHandlers(Vertx vertex,
   @Test
   public void cannotCheckoutWithInvalidCommandCode(Vertx vertex,
       VertxTestContext testContext) throws Throwable {
-    callService("blablabalb", testContext, vertex, result -> {
+      callService("blablabalb", testContext, vertex, result -> {
```
Comment (mreno-EBSCO) [inline src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java]: I don't think this indentation is correct. It shouldn't be at the same level as the lamda block on the next line nor should it line up with the method argument list the line above. I think the original indentation is correct. Is checkstyle requiring this indentation?

File: src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java [hunk 4]
```diff
@@ -79,8 +81,38 @@ public void canStartMainVericleInjectingSip2RequestHandlers(Vertx vertex,
...
   }
+
+  @Test
+  public void canMakeValidSCStatusRequest(Vertx vertex,
+                                          VertxTestContext testContext) throws Throwable {
+
+    callService("9900401.00AY1AZFCA5",
+      testContext, vertex, result -> {
+        String expectedPreLocalTime = "98YYNYNN53" + getFormattedDateString();
+        String expectedPostLocalTime = "1.23|AOfs00000010test|AMChalmers|BXYNNNYNYNNNNNNNYN|ANTL01|AFscreenMessages|AGline\n\n";
```
Comment (mreno-EBSCO) [inline src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java]: Is this an actual response or is something else supposed to happen to it? `AGline\n\n` can't be right? It'd have to be `AGline|\r` to be correct, right?

File: src/test/java/org/folio/edge/sip2/handlers/SCStatusHandlerTests.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    SCStatus.SCStatusBuilder statusBuilder = SCStatus.builder();
+    statusBuilder.maxPrintWidth(20);
+    statusBuilder.protocolVersion("1.00");
+    statusBuilder.statusCode(StatusCode.SC_OK);
+    SCStatus status =  statusBuilder.build();
+
+    String sipMessage = handler.execute(status);
+    //Because the sipMessage has a dateTime component that's supposed to be current, we can't assert on the entirety of the string, have to break it up into pieces.
+    String expectedPreLocalTime = "98YYNYNN53" + getFormattedDateString();
+    String expectedPostLocalTime = "1.23|AOfs00000010test|AMChalmers|BXYNNNYNYNNNNNNNYN|ANTL01|AFscreenMessages|AGline\n\n";
```
Comment (mreno-EBSCO) [inline src/test/java/org/folio/edge/sip2/handlers/SCStatusHandlerTests.java]: See my other comment about the validity of this message.

File: src/main/java/org/folio/edge/sip2/domain/messages/responses/ACSStatus.java [hunk 6]
```diff
@@ -88,7 +88,7 @@ private ACSStatus(ACSStatusBuilder builder) {
     this.protocolVersion = builder.protocolVersion;
     this.institutionId = builder.institutionId;
     this.libraryName = builder.libraryName;
-    this.supportedMessages = Collections.unmodifiableSet(
+    this.supportedMessages =  Collections.unmodifiableSet(
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/edge/sip2/domain/messages/responses/ACSStatus.java]: I think this is some sort of editor fail here. It seems that we are putting extra spaces when dealing with `supportedMessages`. We can probably revert these changes. I'm surprised checkstyle didn't flip out here.

**Comment by mreno-EBSCO** [comment]
No need to revert the enum files. Not having a semicolon is perfectly valid. I actually typed it in and Eclipse doesn't care either way, so it won't try to add it back on or something weird like that (format wars). There is a file with some spaces added, I think incorrectly, that could/should be reverted. You'll see it in the review comments.

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+   * @param command the Command that serves as a Key to the runtime cache
+   * @return Compiled freemarker template
+   */
+  public Template getFreemarkerTemplate(Command command) {
+    return templates.get(command);
+  }
+
+  private FreemarkerRepository() {
+    initializeTemplates();
+    log = LogManager.getLogger(MethodHandles.lookup().lookupClass());
```
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java]: Good catch about the NPE. As for creating the logger in the constructor, I think that's the right place for it, unless you think there's a better place, maybe inline at variable declaration?

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java (hunk 1 above)
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java]: True! So much for my desire to have loosely coupled classes and I overlooked this part of the code!!

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java]: Typically, the pattern is to declare it as a private static final as in:
`private static final Logger log = LogManager.getLogger();`

Since these are one instance classes, I am fine with them being non-static, but I'd declare them at object creation like:
`private final Logger log = LogManager.getLogger();`

I've not seen a logger created in the constructor before, since logging is typically outside the scope of the business logic of the class. Some frameworks inject loggers so you don't have to worry about it.

[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java]: Yeah, i think part of my motivation for instantiating it in the constructor is to be able to inject it in for testing purposes, but I thought that we could mock it via Mockito somehow. I briefly looked into this over the weekend but not able to get Mockito to mock it.  And I didn't want to go overboard with the testing for now, so we can revisit it later.

[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java]: You want to mock the logger?

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java (hunk 2 above)
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java]: debug()'s signature does not accept lambdas.

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java]: if you want 100% coverage or to verify with certainty every line of code! :)  I don't think I want to do that now. given that some negative test cases already run by these logging statements without verifying them, and that we have bigger fish to fry.

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java (hunk 2 above)
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java]: I believe it is supported via the `debug(String message, Supplier<?>... paramSuppliers`.

[Example](https://logging.apache.org/log4j/2.0/manual/api.html#Java_8_lambda_support_for_lazy_logging)

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java]: But we don't make decisions based on what we log, do we? As long as the test gets into the block with the log line, the line is covered, at least that is my understanding. I assume something else more tangible will happen in addition to logging, like throwing an exception or returning `null` that can be checked in the test.

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java (hunk 2 above)
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java]: The braces after ()-> threw it off. I removed the braces and it's working now.

Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java]: Sorry, I thought they might be needed since these were ternary operations, so I added them. I guess I was wrong. 😄

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java]: We don't, but sometimes an error message is critical for troubleshooting, for ease of fishing it out of the logs, so we'd like to make sure that it's there.

File: src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java (hunk 4 above)
Comment (mtraneis) [reply src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java]: I can only get the template to produce "\n" by virtue of  default carriage return. (\r is used in Mac OS environment. \n is Linux, \r\n is Windows.  Hope we don't have to manually create the template on a Mac :))

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+import java.io.Writer;
+import java.lang.invoke.MethodHandles;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+public class FreemarkerUtils {
+
+  private static Logger log;
```
[the remark this answers is not in the document]
Comment (mtraneis) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerUtils.java]: I need to declare the logger outside of the static block to be able to access it globally within the class.  -Resolved-

File: src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mreno-EBSCO) [reply src/main/java/org/folio/edge/sip2/handlers/freemarker/FreemarkerRepository.java]: Sure, but I am not convinced we need to do things that are contrary to best practices in order to ensure messages, that could be suppressed/disable via configuration, are exactly what we expect. I think we should use something built into log4j2, like a plugin: https://stackoverflow.com/questions/24205093/how-to-create-a-custom-appender-in-log4j2

This way, we add it for the test, get the log4j2 messages in the appender plugin and assert they are what we expect.

Here is how someone did that for log4j1.x: https://stackoverflow.com/questions/1827677/how-to-do-a-junit-assert-on-a-message-in-a-logger

**Comment by mreno-EBSCO** [comment]
Oh, wait this is already merged. 😃 I'll stop with the followup comments. We can address them, if we need to, as we go. Feel free to delete the branch.

**Comment by mtraneis** [comment]
> Oh, wait this is already merged. 😃 I'll stop with the followup comments. We can address them, if we need to, as we go. Feel free to delete the branch.

I'll delete the branch but keep on commenting. I can incorporate the changes in the next PR, if they are trivial.
